syntax="proto3";

package pb.store;

option java_package = "org.dingo.store";
option cc_generic_services = true;


enum RegionStatus {
    NONE = 0;
    IDLE = 1;
    DOING = 2;
}

message Range {
    bytes start_key               = 1;
    bytes end_key                 = 2;
}

message RegionInfo {
    uint64 region_id              = 1;
    uint64 region_epoch           = 2;
    uint64 table_id               = 3;
    string table_name             = 4;
    uint64 partition_id           = 5;
    uint32 replica_num            = 6;
    Range range                   = 7;

    repeated string peers         = 10;
    string leader                 = 11;

    RegionStatus status           = 20;

    uint32 create_time            = 30;
};

message AddRegionRequest {
    RegionInfo region_info        = 1;
}

message AddRegionResponse {
}

message DestroyRegionRequest {
    uint64 region_id               = 1;
}

message DestroyRegionResponse {
}

message KvGetRequest {
    uint64 region_id                = 1;
    bytes key                       = 2;
}

message KvGetResponse {
    bytes value                     = 2;
}

message KvPutRequest {
    uint64 region_id                = 1;
    bytes key                       = 2;
    bytes values                    = 3;
}

message KvPutResponse {
}

service StoreService {
    // region
    rpc AddRegion(AddRegionRequest) returns(AddRegionResponse);
    rpc DestroyRegion(DestroyRegionRequest) returns(DestroyRegionResponse);

    // kv
    rpc KvGet(KvGetRequest) returns(KvGetResponse);
    rpc KvPut(KvPutRequest) returns(KvPutResponse);
};
